MAINT: cleanup regionprop add PYTHONOPTIMIZE=2 to travis array

sciunto opened the pull request:
## Description

Normally, this is now solved in scipy, so it should pass here as well.

Fixes #3763


## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

hmaarrfk commented:
These seem to be real errors in the build and should probably be debugged locally.

jni commented:
Looks like it's pyamg that doesn't support PYTHONOPTIMIZE=2 now. I think all the [errors](https://travis-ci.org/scikit-image/scikit-image/jobs/537235136#L2375) are repeats of this:

```pytb
../../venv/lib/python3.5/site-packages/skimage/feature/__init__.py:10: in <module>
    from .peak import peak_local_max
../../venv/lib/python3.5/site-packages/skimage/feature/peak.py:3: in <module>
    from ..segmentation import relabel_sequential
../../venv/lib/python3.5/site-packages/skimage/segmentation/__init__.py:1: in <module>
    from .random_walker_segmentation import random_walker
../../venv/lib/python3.5/site-packages/skimage/segmentation/random_walker_segmentation.py:37: in <module>
    from pyamg import ruge_stuben_solver
../../venv/lib/python3.5/site-packages/pyamg/__init__.py:14: in <module>
    from .gallery import demo
../../venv/lib/python3.5/site-packages/pyamg/gallery/__init__.py:7: in <module>
    from .example import *
../../venv/lib/python3.5/site-packages/pyamg/gallery/example.py:49: in <module>
    load_example.__doc__ %= ('\n' + ' ' * 8).join([''] + example_names)
E   TypeError: unsupported operand type(s) for %=: 'NoneType' and 'str'
```

So @sciunto maybe try moving the PYTHONOPTIMIZE=2 to a build without optional dependencies?

jni commented:
Damn... Looks like not having this on for so long means we've introduced a whole bunch of other errors. =(

Sure makes you appreciate our continuous testing... Imagine how much stuff would continuously break if we didn't have it!

sciunto commented:
I don't think it comes from code we introduced. In this build, scipy is 0.17, not recent enough.
https://travis-ci.org/scikit-image/scikit-image/jobs/537355019#L1009

I added the optimize flag on an entry with min_requirements = 1. My mistake.  I added a new entry, no other choice.

jni commented:
@sciunto perhaps `OPTIONAL_REQUIREMENTS` could be removed from another entry? Sorry to be a pain, but I think our Travis allotment gets hammered enough as it is...

sciunto commented:
Yes, I agree.

sciunto commented:
tests are failing after  ```QtWidgets.QApplication([])```. Qt might not support this flag... :(

hmaarrfk commented:
maybe we can disable QT for that build too?

jni commented:
Wait I take it back it's not passing. :sob:

hmaarrfk commented:
Decorators in python seem really problematic.

https://hynek.me/articles/decorators/

stefanv commented:
I'm not sure what to take away there.  It looks like, under 3.5+, you get mostly reasonable behavior, at least from the user perspective.

hmaarrfk commented:
Tests are failing for all versions, even without optional dependencies

See failure on python 3.7 for example
https://travis-ci.org/scikit-image/scikit-image/jobs/544015869#L2399

This seems to be due to a bug in the python core library.

jni commented:
> This seems to be due to a bug in the python core library

What makes you say that? To me it looks like whatever fancy logic we're using in regionprops doesn't survive PYTHONOPTIMIZE? Having said that, I have approximately 0 understanding of the problems with PYTHONOPTIMIZE.

hmaarrfk commented:
I had some minimal reproducible code that cause the same symptom without any scikit image code. Will post later

sciunto commented:
@hmaarrfk Have you searched for a reported issue to python, as you suggested that the problem takes its roots there?

hmaarrfk commented:
```

    def __getitem__(self, key):
        value = getattr(self, key, None)
        if value is not None:
            return value
        else:  # backwards compatibility
            return getattr(self, PROPS[key])
```

I think this stuff here is causing some real headaches. I'm not sure was trying to be accomplished here.

Seems like some hack around the `@property` decorator, but i'm not sure.

hmaarrfk commented:
The issue is that youa re trying to use the properites as properties and functions.

This is likely clashing with python.

You should wait til 3.8 to make use of 

https://docs.python.org/dev/library/functools.html#functools.cached_property

You know, in 3 years ;)

hmaarrfk commented:
Maybe you can make use of this
https://wiki.python.org/moin/PythonDecoratorLibrary#Cached_Properties

sciunto commented:
I made a test by switching to an optimize level = 1. No change at all.

I discovered that some errors are raised because we have assert calls in the code. I opened another PR (#4006) to go step by step.

hmaarrfk commented:
https://pypi.org/project/cached-property/

sciunto commented:
> ```
> 
>     def __getitem__(self, key):
>         value = getattr(self, key, None)
>         if value is not None:
>             return value
>         else:  # backwards compatibility
>             return getattr(self, PROPS[key])
> ```
> 


I think the backward compatibility is useless as far as all the attributes are defined...

hmaarrfk commented:
I think the property decorator should probably be used, instead of switching functions to properties in some obscure doc mangling function.

sciunto commented:
Let's wait #4006 to be merged, rebase this branch and hopefully, the last failure will go off.

sciunto commented:
@scikit-image/core We need reviews here. This PR adds travis entries with PYTHONOPTIMIZE=2.

hmaarrfk commented:
I think we should probably rebase and force push. This ended up being a two line change.

We can likely remove entries from the build matrix too.

hmaarrfk commented:
An other cleanup you cna do is to make everything external a property
https://github.com/sciunto/scikit-image/pull/5

sciunto commented:
I forgot to fetch before rebasing :man_facepalming:

hmaarrfk commented:
Did you lose your work?

hmaarrfk commented:
https://github.com/sciunto/scikit-image/pull/6

sciunto commented:
Thanks @hmaarrfk

hmaarrfk commented:
You probably want to cleanup the huge build matrix too.

hmaarrfk commented on skimage/measure/_regionprops.py:
```diff
@@ -883,7 +919,6 @@ def _install_properties_docs():
     for p in [member for member in dir(RegionProperties)
               if not member.startswith('_')]:
         getattr(RegionProperties, p).__doc__ = prop_doc[p]
-        setattr(RegionProperties, p, property(getattr(RegionProperties, p)))
```
@scikit-image/core this is the main controbiution of this PR. There seem to be quite a few people interested in region props PRs at this time, so it might be nice to get this one through!

jni commented:
@hmaarrfk I think #4095 provides all of this is well, so I'm wondering whether we should just add the build matrix entry there, rather than force a painful rebase/merge?

hmaarrfk commented:
They seem ok with git.

This has been ready for a while. Let's merge this, release 0.16 and get on with our days.

jni commented:
You made a persuasive argument, @hmaarrfk. ;)
